### DISCUSSION TITLE
Request for Guidance on Training YOLO11n and Addressing Specific Issues

### DISCUSSION THREAD

**Discussion by Brayan532** [ultralytics/ultralytics#17804] [Q&A] [answered]
Dear Glenn,
First of all, I want to sincerely thank you for actively addressing the numerous questions from the community. Your expertise and dedication are truly invaluable.

I am currently working with a lightly modified version of YOLO11n, which I have trained on the COCO dataset (approximately 117k images for training and 5k for validation). However, I have encountered a few challenges that I hope you can help me address with your professional insight. Here are the details:
1. **Training and Transfer Learning Challenges**:
I am using an 8GB GPU, which limits the maximum number of epochs I can run to 60 (as it takes about 24 hours to complete).
In my first round of training, I slightly modified the default configurations and used the following settings: 
_lr0=0.001, warmup=5, mixup=0.2, cos_lr=True_  
The result was a mAP50-95 of 25.6%.
Next, I used transfer learning from last.pt and continued training with updated settings: 
_lr0=0.0005, warmup=7, mixup=0.1, cos_lr=True_  
However, I observed a decline in mAP50-95 after the earlier epochs, which continued until the end of training.
Given this situation, I am uncertain about the best course of action to take next. I see two possible options:
Option 1: Restart training from scratch with new settings.
Option 2: Use transfer learning again from the first training attempt and apply updated settings to save time.
I would greatly appreciate your advice on how best to proceed to achieve better performance.
2. **Validation Dataset Issue**:
During training, I noticed that the validation process only involves 157 items, despite the validation folder containing 5,000 images (as confirmed by the following log):
 _\val\labels.cache... 4952 images, 48 backgrounds, 0 corrupt: 100% 5000/5000 [00:00<?, ?it/s]_  
For reference, my batch size is 16, and I have already cleared the cache and rerun the code, but the issue persists. It seems that while the total number of validation images is recognized correctly, only half of them are being used during validation.
Could you please help me understand why this is happening and how I can ensure the full validation dataset is utilized?
3. **Learning Rate Confusion**:
While training with the following command:
_model.train(data="coco.yaml", epochs=60, imgsz=640, lr0=0.001, cos_lr=True, warmup_epochs=5.0, mixup=0.2)_  
I received the following message in the logs:
_optimizer: 'optimizer=auto' found, ignoring 'lr0=0.001' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically..._  
_optimizer: SGD(lr=0.01, momentum=0.9) with parameter groups 82 weight(decay=0.0), 89 weight(decay=0.0005), 88 bias(decay=0.0)_  
This seems to indicate that my lr0=0.001 setting was overridden, and lr0=0.01 was used instead. However, I noticed in args.yaml (saved in run/detect/train), the value of lr0 is still set to 0.001.
Could you clarify whether the learning rate in the logs (0.01) is the actual value being used, or if the value in args.yaml (0.001) is correct? This inconsistency has been confusing for me.
As someone with significantly more experience, I truly value your insights and guidance. Your advice would be incredibly helpful in resolving these issues and improving my training process.
Thank you very much for your time and support.

Best regards.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Brayan532, thank you for reaching out and for sharing such detailed information about your experiences with YOLO11n 🚀! Your commitment to optimizing your training process is evident, and we're here to help guide you.

This is an automated response to assist you quickly 🎉. Rest assured, an Ultralytics engineer will review your query and provide personalized insights soon.

In the meantime, here are some steps and resources that may help:

1. **Training and Transfer Learning Challenges**:
   - For the issues with mAP decline, you may want to ensure you're following our [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/), which include strategies for learning rate schedules, dataset augmentation, and other best practices.
   - If this is a 🐛 bug or unexpected behavior, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). This helps us debug your workflow more effectively.

2. **Validation Dataset Issue**:
   - To better understand why only 157 images are being evaluated per validation, double-check your dataset's structure and ensure all your validation labels are correctly formatted and present.
   - Providing a structured MRE (as mentioned above) or sharing snippets of relevant training logs can help identify inconsistencies.

3. **Learning Rate Confusion**:
   - From your description, it appears the optimizer may override explicit parameters (`lr0`, `momentum`) when set to `auto`. The discrepancy in logs vs. `args.yaml` is worth investigating further, so sharing the complete logs and configurations might provide clarity.

## Upgrade

Please ensure you are using the latest version of YOLO, as new updates may address any bugs or inconsistencies in earlier versions. You can upgrade your `ultralytics` package with the following command:

```bash
pip install -U ultralytics
```

Ensure you are operating in an environment with [Python>=3.8](https://www.python.org/) and [PyTorch>=1.8](https://pytorch.org/get-started/locally/), as outlined in our repository's [project requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml).

## Environments

Explore the following environments for training and validation. These setups come pre-configured with the necessary dependencies:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Community & Support

In addition to our direct support, you can connect with our global Ultralytics user community for real-time chat and discussions:
- [Discord](https://ultralytics.com/discord) 🎧
- [Discourse](https://community.ultralytics.com)
- [Subreddit](https://reddit.com/r/ultralytics)

## Status

Lastly, check the status of our CI workflows to verify if all current tests are passing. This ensures stability across all YOLO modes and tasks:
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

Feel free to provide additional details or any updates over the course of your experiments. Thanks again for contributing to the Ultralytics community 💡!

**Comment by glenn-jocher** [comment, marked as answer]
@Brayan532 thank you for reaching out with detailed observations and questions. Here’s guidance for your concerns:

1. This is probably due to point 3 and `warmup_epochs` being enabled.
   
2. It's correct. Validation runs at double the batch size of training. So 157x32≈5024 images.

3. The learning rate displayed in logs reflects the value determined by the `auto` optimizer. For strict control, explicitly set `optimizer` (e.g., SGD/Adam) in your training command to avoid overrides. The learning rate in `args.yaml` may mismatch log settings and primarily acts as a storage reference.

Let us know if you encounter further challenges, and feel free to explore more configuration tips at [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/).

**Reply by Brayan532** [reply]
> @Brayan532 thank you for reaching out with detailed observations and questions. Here’s guidance for your concerns:
> 
>     1. This is probably due to point 3 and `warmup_epochs` being enabled.
> 
>     2. It's correct. Validation runs at double the batch size of training. So 157x32≈5024 images.
> 
>     3. The learning rate displayed in logs reflects the value determined by the `auto` optimizer. For strict control, explicitly set `optimizer` (e.g., SGD/Adam) in your training command to avoid overrides. The learning rate in `args.yaml` may mismatch log settings and primarily acts as a storage reference.
> 
> 
> Let us know if you encounter further challenges, and feel free to explore more configuration tips at [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/).


Dear Glenn and Y-T-G,  

Thank you both for your invaluable advice.  

I would like to ask for further clarification regarding the training process of YOLO11n. Specifically, I have been exploring how to train it effectively, and I am curious about the _strategy_ you employed to train YOLO11n from scratch with random weights.  

Could you kindly share whether you initially trained the model for a high _number of epochs_ (e.g., 1000) or followed a different approach, such as training it for 300 epochs and then utilizing transfer learning (though I assume you may not have)? Additionally, I would appreciate it if you could share the settings you used during the training process.  

For instance, did you use the default _settings_ provided in the _default.yaml_ file, or did you make any modifications to them?  

Additionally, I would like to request some guidance regarding _the expected performance metrics_ during training. For example, if the epochs are set to 1000, what level of mAP50-95 can be expected at intermediate stages, such as around 200 epochs? For instance, would mAP50-95 be around 30% at this stage? I understand this is highly dependent on hardware and settings, but based on your experience with YOLO11 variants, particularly YOLO11n, I believe your insights could provide valuable clues.  

To summarize, my key questions are as follows:  
**1.** What strategy did you use to train YOLO11n from scratch?  
**2.** How many epochs did you consider for the training process?  
**3.** What specific settings did you use for training YOLO11n, and did they differ from the _default.yaml_ configuration?  
**4.** Based on your experience, what performance benchmarks (e.g., mAP50-95) can one anticipate during the training process at different stages?  

Thank you in advance for your time and assistance. I look forward to hearing from you.  

Best regards,

**Reply by glenn-jocher** [reply]
@Brayan532 thank you for your thoughtful questions. While the specifics of YOLO11n's internal training strategy involve proprietary details, you can achieve optimal results by following these general guidelines:

1. Training from scratch typically involves 300-500 epochs with a well-structured dataset and default settings as a starting point.
2. Adjustments to hyperparameters like `lr0`, `momentum`, and `weight_decay` can be made based on validation metrics.
3. Performance metrics such as mAP50-95 vary depending on dataset and settings, but improvements are generally incremental, with significant gains stabilizing past early epochs.

For further insights, please refer to the [Train Documentation](https://docs.ultralytics.com/modes/train/). Let us know if you have additional questions!

**Reply by Brayan532** [reply]
Dear Glenn, 
Thank you so much for your swift and constructive advice. 
Best wishes.

**Reply by glenn-jocher** [reply]
You're very welcome! If you have any further questions or need additional guidance, feel free to ask. Best of luck with your YOLO11 training!
